rfcs/20221124-monorepo.md

```diff
@@ -0,0 +1,179 @@
+↖️ Table of Contents
+
+# RFC: Monorepo
+
+**Status:** 🚧 WIP, comments are welcome nonetheless
+
+## Reviewers
+
+- [ ] &lt;your name here&gt;
+- [ ] &lt;your name here&gt;
+
+## Rationale
+
+For a comprehensive discussion about the pros, the cons and the counterpoints to each see [discussion](https://github.com/WordPress/openverse/issues/192). This is not the purpose of this RFC.
+
+This RFC summarily lists the benefits and then, with the twin assumptions of a monorepo being ultimately beneficial and the decision to migrate being finalised in the above discussion, proceeds to go into the implementation details.
+
+### Exclusive benefits of monorepo
+
+This only includes things that cannot be accomplished without the use of a monorepo.
+
+1. Single place to go for issues, PRs and all activity. Currently tickets are scattered across several repos, and any tickets that could benefit more than a single layer must be opened in each of the different repos.
+
+1. Singular copy (different from synced independent copies) of scaffolding code such as Git hooks, lint rules and common workflows. This is distinctly better than elaborate sync workflows.
+
+1. Central place for all technical documentation, enabling documentation for different parts of the stack to cross-reference other pieces and stay current with changes in other places.
+
+1. Enables the infra to deploy the code to coexist with the code itself. Apart from the private secrets that will still need to be encrypted, the IaC files could be organised identical to the code.
+
+1. Milestones that can span across multiple layers of the stack are only possible in GitHub. This is a limitation imposed by GitHub and there is no workaround for this.
+
+The [integration](#step-4-integration) section in the latter part of the document describes more interesting outcomes made possible by the monorepo. They may or may not be exclusive to monorepos but they're surely made easier by it.
+
+## Migration path
+
+First we will merge the API and the frontend. This decision was made for the following reasons.
```

Comment:
This reasoning makes a lot of sense to me, and I'm sold on merging API/frontend first and separately from the Catalog. How much more difficult do you anticipate eventually merging the Catalog to be? Do you think there's value in planning that work first, or can that be tabled until after the API/frontend are merged?

Reply:
I am fairly confident that merging the catalog would not cause major pain later on, mostly because it's another Python project like the API and ingestion server.

We could definitely plan for the eventual catalog-API merge, but I abstained from that in the RFC because
- I am not the most experienced with the catalog so I felt a bit underqualified to write about that in the RFC
- I didn't want to delay the RFC till I could understand the nuances of the catalog enough to include the portion about merging it